unified-scheduler-pool/src/lib.rs

```diff
@@ -687,22 +777,32 @@ impl<S: SpawnableScheduler<TH>, TH: TaskHandler> ThreadManager<S, TH> {
 
         let handler_main_loop = || {
             let pool = self.pool.clone();
-            let mut runnable_task_receiver = runnable_task_receiver.clone();
-            let finished_task_sender = finished_task_sender.clone();
+            let mut blocked_task_receiver = blocked_task_receiver.clone();
+            let mut idle_task_receiver = idle_task_receiver.clone();
+            let finished_blocked_task_sender = finished_blocked_task_sender.clone();
+            let finished_idle_task_sender = finished_idle_task_sender.clone();
 
             move || loop {
                 let (task, sender) = select! {
-                    recv(runnable_task_receiver.for_select()) -> message => {
-                        if let Some(task) = runnable_task_receiver.after_select(message.unwrap()) {
-                            (task, &finished_task_sender)
+                    recv(blocked_task_receiver.for_select()) -> message => {
+                        if let Some(task) = blocked_task_receiver.after_select(message.unwrap()) {
+                            (task, &finished_blocked_task_sender)
                         } else {
                             continue;
                         }
                     },
+                    recv(idle_task_receiver) -> task => {
+                        if let Ok(task) = task {
+                            (task, &finished_idle_task_sender)
+                        } else {
+                            idle_task_receiver = never();
+                            continue;
+                        }
+                    },
```

Comment:
I'm not convinced that the multi-channel set up works correctly without `select_biased!`.

Let's say we have something along the lines of:
1. blocked_task_sender => new context
2. idle_task_sender => idle tasks

The idle tasks should be for the new context, but as far as I can tell, there's nothing preventing them from being randomly picked up before the new context in the handler threads.

I do believe this will work with a `select_biased!` call and the proper ordering, but with the random `select`ing it seems like there's random chance of the chained channel stuff getting messed up wrt to the idle tasks.
Maybe I am missing something?

Reply:
You're totally correct. You saved me. This is a race condition...: f02ebd87a2a085d1ba7c18eb0b1315f7fa5c073e, 4e045d7e2ff98471104a41404a3cf582ef0189e1

> I do believe this will work with a select_biased! call and the proper ordering

Well, this won't work even with `select_biased!`... It'll firstly try to receive blocked, then idle. After that, it'll `sched_yield`. Before the `sched_yeild`, the handler thread still could see the idle task for the next context, if it became visible to the thread just after trying to receive blocked and missed to see new context. scheduler thread sends context then idle task successively between the two `try_recv`s.

I think this is the root cause of the mysterious panic, which i only observed once while running against mb...